File: Acwing/First_Level_0_Basic_Algorithm/Second_DataStructure/ds_835_trip_statistics_string/Main.java
package First_Level_0_Basic_Algorithm.Second_DataStructure.ds_835_trip_statistics_string;



import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    static final int N = 100010;
    static int [][] q = new int[N][26];
    static int idx = 0;
    static int [] cnt = new int[N];

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n;
        n = Integer.valueOf(br.readLine());
        while(n-- > 0){
            String [] str;
            str = br.readLine().split(" ");
            if(str[0].equals("I")){
                insert(str[1]);
            }else{
                System.out.println(query(str[1]));
            }
        }
        br.close();

    }

    public static void insert(String s){
        int t = 0;
        for(int i = 0 ; i < s.length() ; ++i){
            int u = s.charAt(i) - 'a';
            if( q[t][u] <= 0) q[t][u] = ++idx;
            t = q[t][u];
        }
        cnt[t]++;
    }

    public static int query(String s){
        int t = 0;
        for(int i = 0 ; i < s.length() ; ++i){
            int u = s.charAt(i) - 'a';
            if(q[t][u] <= 0) return 0;
            t = q[t][u];
        }
        return cnt[t];
    }
}

File: Acwing/First_Level_0_Basic_Algorithm/Second_DataStructure/ds_835_trip_statistics_string/Main.cpp
#include<iostream>
using namespace std;
const int N = 1000010;
int q[N][26] , n , cnt[N] , idx;

void insert(char * s){
    int t = 0;
    for(int i = 0 ; s[i] ; i++){
        int u = s[i] - 'a';
        if(!q[t][u]) q[t][u] = ++ idx;
        t =  q[t][u];
    }
    cnt[t]++;
}

int query(char * s){
    int t = 0 ;
    for(int i = 0 ; s[i] ; i++){
        int u = s[i] - 'a';
        if(!q[t][u]) return 0;
        t = q[t][u];
    }
    return cnt[t];
}
int main(){
    cin >> n;
    char op[2];
    char s[N];
    while(n--){
        cin >> op >> s;
        if( *op == 'I'){
            insert(s);
        }else if( *op == 'Q'){
            cout << query(s) << endl;
        }
    }
}